src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -465,93 +426,165 @@ protected void InitializeReferences()
+            while (true)
+            {
+                runner.InitializeForScan(this, input, startat, false);
+                runner.runtextpos = runtextpos;
 
-                    int stoppos = RightToLeft ? 0 : input.Length;
+                int stoppos = RightToLeft ? 0 : input.Length;
 
-                    Match? match = InternalPerformScan(reuseMatchObject, input, 0, runner, input, returnNullIfQuick: false);
-                    Debug.Assert(match is not null);
+                // We get the Match by calling Scan. 'input' parameter is used to set the Match text which is only relevante if we are using the Run<TState> string
+                // overload, as APIs that call the span overload (like Count) don't require match.Text to be set, so we pass null in that case.
+                Match? match = ScanInternal(reuseMatchObject, input: usingStringOverload ? runner.runtext : null, 0, runner, input, returnNullIfQuick: false);
+                Debug.Assert(match is not null);
 
-                    // if we got a match, then call the callback function with the match and prepare for next iteration.
-                    if (match.Success)
+                // if we got a match, then call the callback function with the match and prepare for next iteration.
+                if (match.Success)
+                {
+                    if (!reuseMatchObject)
                     {
-                        if (!reuseMatchObject)
-                        {
-                            // We're not reusing match objects, so null out our field reference to the instance.
-                            // It'll be recreated the next time one is needed.
-                            runner.runmatch = null;
-                        }
+                        // We're not reusing match objects, so null out our field reference to the instance.
+                        // It'll be recreated the next time one is needed.
+                        runner.runmatch = null;
+                    }
 
-                        if (!callback(ref state, match))
-                        {
-                            // If the callback returns false, we're done.
+                    if (!callback(ref state, match))
+                    {
+                        // If the callback returns false, we're done.
 
-                            if (reuseMatchObject)
-                            {
-                                // We're reusing the single match instance, so clear out its text as well.
-                                // We don't do this if we're not reusing instances, as in that case we're
-                                // dropping the whole reference to the match, and we no longer own the instance
-                                // having handed it out to the callback.
-                                match.Text = null;
-                            }
-                            return;
+                        if (usingStringOverload && reuseMatchObject)
```

Comment:
Assigning null is pretty cheap.  Do we functionally need this usingStringOverload, or could we just always null out match.Text if reuseMatchObject?  If there's negligible perf difference, I'd rather keep it simpler.  Might even be a tad faster by not having the extra argument and branch.

Reply:
Sorry, I missed this comment will send a short follow up addressing this one